File: core/player/prop/trail.js
import { world } from "../../world/world.js";
import { ctx } from "../../world/canvas.js";
import { toroidalDelta } from "../../world/utils.js";

export default class Trail {
    constructor(length, color = "blue") {
        this.length = length;
        this.color = color;
        this.trails = [];
    }

    update(x, y, speed) {
        if (this.trails.length > this.length) {
            this.trails.shift();
        }

        this.trails.push({ x, y, speed });
    }

    render() {
        if (this.trails.length < 2) return;
        for (let i = 1; i < this.trails.length; i++) {
            const a = this.trails[i - 1], b = this.trails[i];
            const avgSpeed = (a.speed + b.speed) / 2;

            const adx = toroidalDelta(world.x, a.x, world.width);
            const ady = toroidalDelta(world.y, a.y, world.height);
            const bdx = toroidalDelta(world.x, b.x, world.width);
            const bdy = toroidalDelta(world.y, b.y, world.height);

            const t = i / this.trails.length;

            if (Math.abs(adx - bdx) > avgSpeed || Math.abs(ady - bdy) > avgSpeed) continue;
            ctx.beginPath();
            ctx.moveTo(adx, ady);
            ctx.lineTo(bdx, bdy);
            ctx.strokeStyle = this.color;

            ctx.globalAlpha = t * 0.1;
            ctx.lineWidth = t * 2;
            ctx.stroke();
            ctx.globalAlpha = 1;
        }
    }
}

File: void_front/game/core/player/prop/trail.py
from __future__ import annotations

from dataclasses import dataclass, field

from game.core.world.canvas import draw_line
from game.core.world.utils import toroidalDelta
from game.core.world.world import world


@dataclass
class Trail:
    length: int
    color: str = "blue"
    trails: list[dict] = field(default_factory=list)

    def update(self, x, y, speed):
        if len(self.trails) > self.length:
            self.trails.pop(0)
        self.trails.append({"x": x, "y": y, "speed": speed})

    def render(self):
        if len(self.trails) < 2:
            return
        for i in range(1, len(self.trails)):
            a = self.trails[i - 1]
            b = self.trails[i]
            avgSpeed = (a["speed"] + b["speed"]) / 2
            adx = toroidalDelta(world.x, a["x"], world.width)
            ady = toroidalDelta(world.y, a["y"], world.height)
            bdx = toroidalDelta(world.x, b["x"], world.width)
            bdy = toroidalDelta(world.y, b["y"], world.height)
            if abs(adx - bdx) > avgSpeed or abs(ady - bdy) > avgSpeed:
                continue
            draw_line((adx, ady), (bdx, bdy), color=self.color, width=max(1, int((i / len(self.trails)) * 2)), alpha=max(0, min(255, int((i / len(self.trails)) * 0.1 * 255))))
